File: src/main/java/com/example/pacman/Ghost.java
package com.example.pacman;
import static com.example.pacman.Location.Direction;
import static com.example.pacman.Location.Direction.*;
import static com.example.pacman.Colours.Colour;
import static com.example.pacman.Colours.Colour.*;

import java.util.Random;

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;


public class Ghost extends GameElement implements Moveable {
    private Direction direction;
    private boolean moving;
    private static final Colour[] COLOURS = {
      RED,
      GREEN,
      PURPLE,
      CYAN
    };
    private static final int SCORE=200;
    private static final int PANIC_LEVEL=50;
    private static final int CALM_LEVEL=0;
    private static int colour=0;
    private Colour normalColour;
    private static final Colour PANIC_COLOUR = BLUE;
    private boolean passedGate=false;
    private Random randomizer = new Random();
    private Behaviour behaviour = new CalmBehaviour();

    public Ghost(Location location) {
        super(location);
        normalColour = COLOURS[colour%COLOURS.length];
        colour++;
        setDirection(UP);
    }

    @Override
    public void tick() {
    	behaviour.tick();
        chooseDirection();
        move();
        checkCollisions();
    }

    @Override
    public void triggerEffect(GameElement element) {
      if (element instanceof PowerPill) {
        panic();
      }
      if (element instanceof Pacman) {
        behaviour.triggerPacmanEffect((Pacman) element);
      }
    }
    
    private void chooseDirection() {
      Direction next = chooseOption(gatherOptions());
      if (next != null) {
        setDirection(next);
      }
    }

    private void move() {
      if (behaviour.shouldMove() && isClear(direction())) {
          setLocation(location().next(direction()));
      }
    }
    
    private void checkCollisions() {
      GameElement pacman = super.getGame().getGameElementByType(Pacman.class);
      if (pacman != null && location().equals(pacman.location())) {
          triggerEffect(pacman);
      }
      GameElement gate = super.getGame().getGameElementByType(Gate.class);
      if (gate != null && location().equals(gate.location()))  {
        passedGate = true;
      }
    }

    private List<Direction> gatherOptions() {
      List<Direction> options = new ArrayList<>();
      List<Direction> possibleRoutes = Arrays.asList(direction(),
                                                      direction().turnLeft(),
                                                      direction().turnRight());
      for (Direction dir: possibleRoutes) {
        if (isClear(dir)) {
          options.add(dir);
        }
      }
      return options;
    }

    private Direction chooseRandom(List<Direction> options){
      return options.get(randomizer.nextInt(options.size()));
    }

    private Direction chooseOption(List<Direction> options) {
      Direction option;
      if (options.isEmpty()) {
        option = behaviour.noChoiceDirection();
      } else {
        option = chooseRandom(options);
      }
      return option;
    }

    public void panic() {
    	behaviour = new PanickedBehaviour();
        setIcon(GhostToken.getToken(PANIC_LEVEL));
    }

    @Override
    public String render() {
      return this.icon();
    }
    
    @Override
    public Colour getColour() {
      return behaviour.getColour();
    }

    @Override
    public void kill() {
      behaviour = new CalmBehaviour();
      setIcon(GhostToken.getToken(CALM_LEVEL));
      restart();
      super.getGame().addScore(SCORE);
    }

    @Override
    public void restart() {
      passedGate = false;
      super.restart();
    }
    
    public boolean isClear(Direction direction) {
      GameElement el = super.getGame().getGameElement(location().next(direction));
      return !(el instanceof Wall || 
               el instanceof ForceField || 
               (el instanceof Gate && passedGate)
              );
    }
    public void setDirection(Direction direction) {
      this.direction = direction;
    }
    public Direction direction() {
      return direction;
    }
    public boolean moving() {
      return moving;
    }
    public void move(Direction direction) {
      // Ghosts are not controlled externally
    }

    public void go() {
      moving = true;
    }

    public void stop() {
      moving = false;
    }
    
    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Ghost) {
            final Ghost ghost = (Ghost) obj;
            return (this.location().equals(ghost.location()));
        } else {
          return false;
        }
    }
    
    @Override
    public int hashCode() {
      return super.hashCode() * 31;
    }
    
    private interface Behaviour {
    	void tick();
    	Colour getColour();
		boolean panicked();
		Direction noChoiceDirection();
    	boolean shouldMove();
    	void triggerPacmanEffect(Pacman pacman);
    	
    }
    
    private class PanickedBehaviour implements Behaviour {

    	private int turnsLeft = PANIC_LEVEL;
    	
		@Override
		public void tick() {
			GameElement pacman = getGame().getGameElementByType(Pacman.class);
	        Location pacmanLocation;
	        pacmanLocation = (pacman!=null) ? pacman.location() : location();
	        setDirection(location().avoid(pacmanLocation));
	        turnsLeft--;
	        if (turnsLeft == 0) {
	        	behaviour = new CalmBehaviour();
	        }
	        setIcon(GhostToken.getToken(turnsLeft));
		}

		@Override
		public Direction noChoiceDirection() {
			return null;
		}

		@Override
		public boolean shouldMove() {
			return turnsLeft%2==0;
		}

		@Override
		public void triggerPacmanEffect(Pacman pacman) {
			kill();
		}

		@Override
		public boolean panicked() {
			return true;
		}

		@Override
		public Colour getColour() {
			return PANIC_COLOUR;
		}
    	
    }
    
    private class CalmBehaviour implements Behaviour {

		@Override
		public void tick() {
			// No-op
		}

		@Override
		public Direction noChoiceDirection() {
			return direction().turnBack();
		}

		@Override
		public boolean shouldMove() {
			return true;
		}

		@Override
		public void triggerPacmanEffect(Pacman pacman) {
			pacman.kill();
		}

		@Override
		public boolean panicked() {
			return false;
		}

		@Override
		public Colour getColour() {
			return normalColour;
		}
    	
    }
}

File: src/main/go/src/pacman/game/ghost.go
package game

import (
	"math/rand"
	"pacman/dir"
)

// Ghost interface extending Element
type Ghost interface {
	Element
	Panic()
	IsPanicked() bool
}

const normalGhost rune = 'M'
const panicGhost rune = 'W'

var panicColour = Colour{BLUE, BLACK}

var ghostIcon = map[rune]bool{
	normalGhost: false,
	panicGhost:  true,
}

const panicLevel int = 50
const ghostPoints int = 200

type ghostStruct struct {
	Element
	panic      int
	origColour Colour
	gatePassed bool
	game       Game
	behaviour  behaviour
}

var ghostColours = []ColourAtt{RED, CYAN, GREEN, PURPLE}
var ghostColour = 0

type behaviour interface {
	tick(*ghostStruct)
	noChoiceDirection(dir.Direction) dir.Direction
	shouldMove() bool
	triggerEffect(Element, *ghostStruct)
}

type panicBehaviour struct {
	turnsLeft int
}

func (p *panicBehaviour) tick(ghost *ghostStruct) {
	pacman := ghost.game.GetPacman()
	if pacman != nil {
		ghost.SetDirection(ghost.Location().Avoid(pacman.Location()))
	}
	ghost.SetColour(panicColour)
	p.turnsLeft--
	if p.turnsLeft == 0 {
		ghost.behaviour = &calmBehaviour{}
		return
	}
}

func (*panicBehaviour) noChoiceDirection(currentDirection dir.Direction) dir.Direction {
	return currentDirection
}

func (p *panicBehaviour) shouldMove() bool {
	return p.turnsLeft%2 == 0
}

func (p *panicBehaviour) triggerEffect(pacman Element, ghost *ghostStruct) {
	ghost.game.SetScore(ghost.game.Score() + ghostPoints)
	ghost.behaviour = &calmBehaviour{}
	ghost.gatePassed = false
	ghost.Restart()
}

type calmBehaviour struct {
}

func (c *calmBehaviour) tick(ghost *ghostStruct) {
	ghost.SetColour(ghost.origColour)
}

func (c *calmBehaviour) shouldMove() bool {
	return true
}

func (c *calmBehaviour) noChoiceDirection(currentDirection dir.Direction) dir.Direction {
	return currentDirection.Opposite()
}

func (c *calmBehaviour) triggerEffect(pacman Element, ghost *ghostStruct) {
	pacman.TriggerEffect(ghost)
}

// NewGhost creates a clean populated ghostStruct
func NewGhost(game Game, icon rune, loc dir.Location) Ghost {
	var colour Colour
	if game != nil {
		colour = Colour{ghostColours[ghostColour], BLACK}
		ghostColour = (ghostColour + 1) % len(ghostColours)
	}
	var startingBehaviour behaviour
	element := NewElement(game, icon, loc, dir.LEFT, ghostPoints)
	if GhostPanic(icon) {
		startingBehaviour = &panicBehaviour{
			turnsLeft: panicLevel,
		}
		element.SetColour(panicColour)
	} else {
		startingBehaviour = &calmBehaviour{}
		element.SetColour(colour)
	}
	return &ghostStruct{Element: element,
		behaviour:  startingBehaviour,
		origColour: colour,
		gatePassed: false,
		game:       game}
}

// Tick activates this elements turn
func (g *ghostStruct) Tick() {
	g.behaviour.tick(g)
	g.chooseDirection()
	g.checkCollisions()
	g.SetIcon(GhostIcon(g.IsPanicked()))
}

func (g *ghostStruct) chooseDirection() {
	ahead := g.Direction()
	choices := []dir.Direction{ahead, ahead.Left(), ahead.Right()}
	options := g.findOptions(choices)
	if options != nil {
		g.SetDirection(randomChoice(options))
	} else {
		g.SetDirection(g.behaviour.noChoiceDirection(ahead))
	}
	g.move()
}

func (g *ghostStruct) findOptions(choices []dir.Direction) []dir.Direction {
	var options []dir.Direction
	for _, nextDir := range choices {
		if g.isClear(g.Location().Next(nextDir)) {
			options = append(options, nextDir)
		}
	}
	return options
}

func randomChoice(options []dir.Direction) dir.Direction {
	index := rand.Intn(len(options))
	return options[index]
}

func (g *ghostStruct) move() {
	if g.behaviour.shouldMove() && g.isClear(g.Location().Next(g.Direction())) {
		g.SetLocation(g.Location().Next(g.Direction()))
	}
}

func (g *ghostStruct) checkCollisions() {
	gate := g.game.GetGate()
	if gate != nil && g.Location() == gate.Location() {
		g.gatePassed = true
	}
	pacman := g.game.GetPacman()
	if pacman != nil && g.Location() == pacman.Location() {
		g.TriggerEffect(pacman)
	}
}

// Panic the ghost
func (g *ghostStruct) Panic() {
	g.behaviour = &panicBehaviour{
		turnsLeft: panicLevel,
	}
	g.SetIcon(GhostIcon(true))
	g.SetColour(panicColour)
}

// IsPanicked returns true is the ghost panic level is non-zero
func (g *ghostStruct) IsPanicked() bool {
	_, panicking := g.behaviour.(*panicBehaviour)
	return panicking
}

// TriggerEffect of colliding with this element
func (g *ghostStruct) TriggerEffect(pacman Element) {
	g.behaviour.triggerEffect(pacman, g)
}

func (g *ghostStruct) isClear(nextLoc dir.Location) bool {
	clear := true
	walls := g.game.GetWalls()
	for _, wall := range walls {
		if wall.Location() == nextLoc {
			if !wall.IsGate() {
				clear = false
			} else {
				if g.gatePassed {
					clear = false
				}
			}
		}
	}
	return clear
}

// GetGhost returns a new Ghost if the icon is a ghost
func GetGhost(icon rune, location dir.Location) Element {
	if IsGhost(icon) {
		return NewGhost(nil, icon, location)
	}
	return nil
}

// IsGhost returns true is the icon represents one
func IsGhost(icon rune) bool {
	_, ok := ghostIcon[icon]
	return ok
}

// GhostPanic returns true is the icon represents a panicked ghost
func GhostPanic(icon rune) bool {
	panic, _ := ghostIcon[icon]
	return panic
}

// GhostIcon returns a suitable icon for a panicked ghost
func GhostIcon(panic bool) rune {
	if panic {
		return panicGhost
	}
	return normalGhost
}

// AddToGame adds a new type of this element to the game
func (g *ghostStruct) AddToGame(game Game) {
	ghost := NewGhost(game, g.Icon(), g.Location())
	game.AddGhost(ghost)
}
